File: C++/src/PolygonViewer.cpp
#include "PolygonViewer.hpp"

namespace PolygonViewerLibrary {

  void PolygonViewer::ViewPolygon(const string& fileName) const
  {
    /// Open file
    ofstream ofile;
    ofile.open(fileName);
    if(ofile.fail())
      throw runtime_error("Cannot open file. Something went wrong");

    Polygon initialPolygon = _geometryFactory.GetPolygon();

    /// Data input
    // scrivo i punti che sono i vertici del poligono
    ofile << "points = [";
    for(int i = 0; i < _geometryFactory.GetPolygon().NumberVertices - 1; i++)
      ofile << _geometryFactory.GetPoint(initialPolygon.PolygonVertices[i]) << endl;
    ofile << _geometryFactory.GetPoint(initialPolygon.PolygonVertices[_geometryFactory.GetPolygon().NumberVertices - 1]) << "];" << endl;
    // scrivo i comandi che su matlab permettono di costruire un poligono
    ofile << "polygon = polyshape(points(:, 1),points(:, 2));" << endl;

    ofile << "ptlbls = \"";
    for(int i = 0; i < _geometryFactory.GetPolygon().NumberVertices-1; i++)
        ofile << i << ", ";
    ofile << _geometryFactory.GetPolygon().NumberVertices-1 << '"' << ";" << endl
          << "pointlabels = strsplit(ptlbls, [\",\", \" \"]);"<< endl << endl;


    // scrivo i punti che sono gli estremi del segmento tagliante
    ofile << "segment = [";
    ofile << _geometryFactory.GetPoint(_geometryFactory.GetLine().From) << endl
          << _geometryFactory.GetPoint(_geometryFactory.GetLine().To) << "];" << endl << endl;

    ofile << "seglabels = [ 'P1'; 'P2'];"<<endl;

    /// Functions
    // scrivo i comandi che su matlab permettono di plottare segmenti,poligoni e punti
    ofile << "figure;" << endl << "plot(polygon);" << endl << "hold on;" << endl
          << "plot(segment(:, 1), segment(:, 2), 'r--');" << endl << "hold on;" << endl
          << "plot(points(:, 1),points(:, 2), 'ko');" << endl << "hold on;" << endl
          << "plot(segment(:, 1), segment(:, 2), 'r*');" << endl
          << "text(points(:, 1),points(:, 2), pointlabels', "
          << "'VerticalAlignment','bottom','HorizontalAlignment','right', 'FontSize', 14);" << endl
          << "text(segment(:, 1), segment(:, 2), seglabels, "
          << "'VerticalAlignment','top','HorizontalAlignment','left', 'FontSize', 14, 'Color', 'r');" << endl;


    /// Close file
    ofile.close();
  }

  void PolygonViewer::ViewLine(const string& fileName) const
  {
    /// Create file
    ofstream ofile;
    ofile.open(fileName);
    if(ofile.fail())
      throw runtime_error("Cannot open file. Something went wrong");

    /// Data input
    // scrivo i punti che sono gli estremi del segmento tagliante
    ofile << "segment = [";
    ofile << _geometryFactory.GetPoint(_geometryFactory.GetLine().From) << endl
          << _geometryFactory.GetPoint(_geometryFactory.GetLine().To) << "];" << endl << endl;

    /// Functions
    // scrivo i comandi che su matlab permettono di plottare segmenti e punti
    ofile << "figure;" << endl << "plot(segment(:, 1), segment(:, 2), 'r--');" << endl
          << "hold on;" << endl << "plot(segment(:, 1), segment(:, 2), 'r*');" << endl;

    /// Close file
    ofile.close();
  }

  void PolygonViewer::ViewFinishedProblem(const string &fileName) const
  {
      /// Create file
      ofstream ofile;
      ofile.open(fileName);
      if(ofile.fail())
        throw runtime_error("Cannot open file. Something went wrong");

      ofile << "figure;" << endl;

      ofile << "ptlbls = \"";
      for(int i = 0; i < _geometryFactory.GetPointsNumber()-1; i++)
          ofile << i << ", ";
      ofile << _geometryFactory.GetPointsNumber()-1 << '"' << ";" << endl
            << "pointlabels = strsplit(ptlbls, [\",\", \" \"]);"<< endl
            << "seglabels = [ 'P1'; 'P2' ];"<< endl << endl;

      /// Data input

      ofile << "segment = [";
      ofile << _geometryFactory.GetPoint(_geometryFactory.GetLine().From) << endl
            << _geometryFactory.GetPoint(_geometryFactory.GetLine().To) << "];" << endl << endl;

      unsigned int n = _geometryFactory.GetNumberLeftSubPolygons();
      for(unsigned int i = 0; i < n; i++)
      {
          Polygon polygon = _geometryFactory.GetLeftSubPolygon(i);

          ofile << "points" << i << " = [";
          for(int j = 0; j < polygon.NumberVertices - 1; j++)
            ofile << _geometryFactory.GetPoint(polygon.PolygonVertices[j]) << endl;
          ofile << _geometryFactory.GetPoint(polygon.PolygonVertices[polygon.NumberVertices - 1]) << "];" << endl;

          ofile << "polygon" << i <<" = polyshape(points" << i << "(:, 1),points" << i << "(:, 2));" << endl;

          ofile << "plot(polygon" << i << ");"<< endl << "hold on" << endl << endl;
      }

      for(int i = 0; i < _geometryFactory.GetNumberRightSubPolygons(); i++)
      {
          Polygon polygon = _geometryFactory.GetRightSubPolygon(i);

          ofile << "points" << i + n << " = [";
          for(int j = 0; j < polygon.NumberVertices - 1; j++)
            ofile << _geometryFactory.GetPoint(polygon.PolygonVertices[j]) << endl;
          ofile << _geometryFactory.GetPoint(polygon.PolygonVertices[polygon.NumberVertices - 1]) << "];" << endl;

          ofile << "polygon" << i + n<<" = polyshape(points" << i + n<< "(:, 1),points" << i + n<< "(:, 2));" << endl;
          ofile << "plot(polygon" << i + n << ");"<< endl << "hold on" << endl << endl;
      }

      ofile << endl << "problempoints = [";
      for(int i = 0; i < _geometryFactory.GetPointsNumber(); i++)
        ofile << endl << _geometryFactory.GetPoint(i);
      ofile << "];" << endl << endl;

      ofile << endl << "plot(segment(:, 1), segment(:, 2), 'r--');" << endl
            << "hold on;" << endl << "plot(segment(:, 1), segment(:, 2), 'r*');" << endl<< endl
            << "text(problempoints(:, 1),problempoints(:, 2), pointlabels', "
            << "'VerticalAlignment','bottom','HorizontalAlignment','right', 'FontSize', 14);" << endl
            << "text(segment(:, 1), segment(:, 2), seglabels, "
            << "'VerticalAlignment','top','HorizontalAlignment','left', 'FontSize', 14, 'Color', 'r');" << endl;

      /// Close file
      ofile.close();
  }

}

File: Python/src/polygon_viewer.py
from src import geometry_core as g

### \class PolygonViewer contains methods to visualize the polygon and the line
class PolygonViewer():
    ### \brief construct the reference to GeometryFactory
    def __init__(self, geometryFactory: g.IGeometryFactory):
        self.__geometryFactory = geometryFactory

    ### \brief ViewPolygon() creates a file .m in order to plot the initial polygon and the cutting segment
    def view_polygon(self, file_name: str):
        ### Open file
        try:
           ofile = open(file_name, 'w')
        except IOError:
            raise RuntimeError("Cannot open file. Something went wrong")

        initialPolygon = self.__geometryFactory.GetPolygon()

        ### Data input
        # scrivo i punti che sono i vertici del poligono
        ofile.write("points = [")
        for i in range(0, self.__geometryFactory.GetPolygon().number_vertices - 1):
            ofile.write(str(self.__geometryFactory.GetPoint(initialPolygon.polygon_vertices[i]).x) + ", " +\
                        str(self.__geometryFactory.GetPoint(initialPolygon.polygon_vertices[i]).y) + ";\n")
        ofile.write(str(self.__geometryFactory.GetPoint(initialPolygon.polygon_vertices[self.__geometryFactory.GetPolygon().number_vertices - 1]).x) + ", " +\
                        str(self.__geometryFactory.GetPoint(initialPolygon.polygon_vertices[self.__geometryFactory.GetPolygon().number_vertices - 1]).y) + "];\n")
        # scrivo i comandi che su matlab permettono di costruire un poligono
        ofile.write("polygon = polyshape(points(:, 1),points(:, 2));\n")
        # scrivo i punti che sono gli estremi del segmento tagliante
        ofile.write("segment = [")
        ofile.write(str(self.__geometryFactory.GetPoint(self.__geometryFactory.GetLine().From).x) + ", " +\
                    str(self.__geometryFactory.GetPoint(self.__geometryFactory.GetLine().From).y) + ";\n" + \
                    str(self.__geometryFactory.GetPoint(self.__geometryFactory.GetLine().To).x) + ", " + \
                    str(self.__geometryFactory.GetPoint(self.__geometryFactory.GetLine().To).y) + "];\n\n")

        ### Functions
        # scrivo i comandi che su matlab permettono di plottare segmenti,poligoni e punti
        ofile.write("figure;" + "\n" + "plot(polygon);" + "\n" + "hold on;" + "\n" +\
                    "plot(segment(:, 1), segment(:, 2), 'r--');" + "\n" + "hold on;" + "\n" +\
                    "plot(points(:, 1),points(:, 2), 'ko');" + "\n" + "hold on;" + "\n" +\
                    "plot(segment(:, 1), segment(:, 2), 'r*');" + "\n")

        ### Close file
        ofile.close()

    ### \brief ViewLine() creates a file .m in order to plot the cutting segment
    def view_line(self, file_name: str):
        ### Open file
        try:
            ofile = open(file_name, 'w')
        except IOError:
            raise RuntimeError("Cannot open file. Something went wrong")

        ### Data input
        # scrivo i punti che sono gli estremi del segmento tagliante
        ofile.write("segment = [")
        ofile.write(str(self.__geometryFactory.GetPoint(self.__geometryFactory.GetLine().From).x) + ", " +\
                    str(self.__geometryFactory.GetPoint(self.__geometryFactory.GetLine().From).y) + ";\n" + \
                    str(self.__geometryFactory.GetPoint(self.__geometryFactory.GetLine().To).x) + ", " + \
                    str(self.__geometryFactory.GetPoint(self.__geometryFactory.GetLine().To).y) + "];\n\n")

        ### Functions
        # scrivo i comandi che su matlab permettono di plottare segmenti e punti
        ofile.write("figure;" + "\n" + "plot(segment(:, 1), segment(:, 2), 'r--');" + "\n" +\
                    "hold on;" + "\n" + "plot(segment(:, 1), segment(:, 2), 'r*');" + "\n")

        ### Close file
        ofile.close()
